service-clients/src/main/java/com/google/cloud/pso/rag/common/JDBCHelper.java

```diff
@@ -0,0 +1,92 @@
+package com.google.cloud.pso.rag.common;
+
+import com.pgvector.PGvector;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class JDBCHelper {
+  private static final ExecutorService databaseExecutor = Executors.newFixedThreadPool(1);
+
+  public static CompletableFuture<ResultSet> executeSQLStmtAsync(
+      String sql, String jdbcUrl, String user, String password) {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          // JDBC operations happen in a thread from databaseExecutor
+          try (Connection connection = DriverManager.getConnection(jdbcUrl, user, password);
+              Statement stmt = connection.createStatement()) {
+            stmt.execute(sql);
+            return stmt.getResultSet();
+          } catch (SQLException e) {
+            throw new RuntimeException("Database error during query: " + e.getMessage(), e);
+          }
+        },
+        databaseExecutor);
+  }
+
+  public static <T> Stream<T> streamResultSet(ResultSet rs, ResultSetMapper<T> mapper)
```

Comment:
maybe using a library can be of help here: 
- jooq does have stream support along with other niceties -> https://www.jooq.org/doc/latest/manual/sql-execution/fetching/lazy-fetching-with-streams/
- a more advance but also modern approach could be implemented using R2DBC implementations -> https://github.com/pgjdbc/r2dbc-postgresql